stats/statsvariables.cpp

```diff
@@ -0,0 +1,1674 @@
+	bool preferBin(const StatsBin &bin) const override {
+		year_quarter value = derived_bin(bin).value;
+		return value.second == 1;
+	}
+	year_quarter to_bin_value(const dive *d) const {
+		struct tm tm;
+		utc_mkdate(d->when, &tm);
+
+		int year = tm.tm_year;
+		switch (tm.tm_mon) {
+		case 0 ... 2: return { year, 1 };
+		case 3 ... 5: return { year, 2 };
+		case 6 ... 8: return { year, 3 };
+		default:      return { year, 4 };
+		}
+	}
+	void inc(DateQuarterBin &bin) const {
+		if (++bin.value.second > 4) {
+			bin.value.second = 1;
+			++bin.value.first;
+		}
+	}
+};
+
+using DateMonthBin = SimpleBin<year_month>;
+
+struct DateMonthBinner : public SimpleContinuousBinner<DateMonthBinner, DateMonthBin> {
+	QString name() const override {
+		return StatsTranslations::tr("Monthly");
+	}
+	QString format(const StatsBin &bin) const override {
+		year_month value = derived_bin(bin).value;
+		return QString("%1 %2").arg(monthname(value.second), QString::number(value.first));
+	}
+	// In histograms, output year for fill years, month otherwise
+	QString formatLowerBound(const StatsBin &bin) const override {
+		year_month value = derived_bin(bin).value;
+		return value.second == 0 ? QString::number(value.first)
+					 : QString(monthname(value.second));
+	}
+	double lowerBoundToFloatBase(year_quarter value) const {
+		return date_to_double(value.first, value.second, 0);
+	}
+	// Prefer bins that show full years
+	bool preferBin(const StatsBin &bin) const override {
+		year_month value = derived_bin(bin).value;
+		return value.second == 0;
+	}
+	year_month to_bin_value(const dive *d) const {
+		struct tm tm;
+		utc_mkdate(d->when, &tm);
+		return { tm.tm_year, tm.tm_mon };
+	}
+	void inc(DateMonthBin &bin) const {
+		if (++bin.value.second > 11) {
+			bin.value.second = 0;
+			++bin.value.first;
+		}
+	}
+};
+
+static DateYearBinner date_year_binner;
+static DateQuarterBinner date_quarter_binner;
+static DateMonthBinner date_month_binner;
+struct DateType : public StatsVariableTemplate<StatsVariable::Type::Continuous> {
+	QString name() const {
+		return StatsTranslations::tr("Date");
+	}
+	double toFloat(const dive *d) const override {
+		return d->when / 86400.0;
+	}
+	std::vector<const StatsBinner *> binners() const override {
+		return { &date_year_binner, &date_quarter_binner, &date_month_binner };
+	}
+};
+
+// ============ Dive depth, binned in 5, 10, 20 m or 15, 30, 60 ft bins ============
+
+struct DepthBinner : public IntRangeBinner<DepthBinner, IntBin> {
+	bool metric;
+	DepthBinner(int bin_size, bool metric) : IntRangeBinner(bin_size), metric(metric)
+	{
+	}
+	QString name() const override {
+		QLocale loc;
+		return StatsTranslations::tr("in %1 %2 steps").arg(loc.toString(bin_size),
+								   get_depth_unit());
+	}
+	QString unitSymbol() const override {
+		return get_depth_unit();
+	}
+	int to_bin_value(const dive *d) const {
+		return metric ? d->maxdepth.mm / 1000 / bin_size
+			      : lrint(mm_to_feet(d->maxdepth.mm)) / bin_size;
```

Comment:
I guess this uses the preference, not a variable... but it seemed that I was never given an option to use anything but what's used in the preferences, so...
https://github.com/subsurface/subsurface/blob/edd19d4725153e6428e7b2e1b6594b4168601a39/core/units.h#L27

Reply:
This macro does something completely different, no? It is used of you want to provide two "nice" numbers.

In any case - the code was very inconsistent : Sometimes it would respect the "metric" variable, sometimes use the preferences flag. After some thinking I concluded that the correct thing to do is using the variable. For now, this is an obscure difference. But what if we save the statistics into presets similar to the filters? The user may end up with loading a binner for a different unit system than in the preferences. Therefore the binners now all format according to their proper type, just to be sure. The numeric representations don't have that problem and therefore use the preferences flag.

Reply:
Yes, that's a good point. The ability to have bins in a different unit system than the one currently set in preferences might avoid pain later.